packages/android_intent/android/src/main/java/dev/flutter/plugins/androidintent/IntentSender.java

```diff
@@ -0,0 +1,116 @@
+  }
+
+  /**
+   * Creates and launches an intent with the given params using the cached {@link Activity} and
+   * {@link Context}.
+   *
+   * <p>This will fail to create and send the intent if {@code applicationContext} hasn't been set
+   * at the time of calling.
+   *
+   * <p>This uses {@code activity} to start the intent whenever it's not null. Otherwise it falls
+   * back to {@code applicationContext} and adds {@link Intent#FLAG_ACTIVITY_NEW_TASK} to the intent
+   * before launching it.
+   *
+   * @param action the Intent action, such as {@code ACTION_VIEW}.
+   * @param flags forwarded to {@link Intent#addFlags(int)} if non-null.
+   * @param category forwarded to {@link Intent#addCategory(String)} if non-null.
+   * @param data forwarded to {@link Intent#setData(Uri)} if non-null.
+   * @param arguments forwarded to {@link Intent#putExtras(Bundle)} if non-null.
+   * @param packageName forwarded to {@link Intent#setPackage(String)} if non-null. This is forced
+   *     to null if it can't be resolved.
+   * @param componentName forwarded to {@link Intent#setComponent(ComponentName)} if non-null.
+   */
+  void send(
+      String action,
+      @Nullable Integer flags,
+      @Nullable String category,
+      @Nullable Uri data,
+      @Nullable Bundle arguments,
+      @Nullable String packageName,
+      @Nullable ComponentName componentName) {
+    if (applicationContext == null) {
+      Log.wtf(TAG, "Trying to send an intent before the applicationContext was initialized.");
+      return;
+    }
+
+    Intent intent = new Intent(action);
+
+    if (flags != null) {
+      intent.addFlags(flags);
+    }
+    if (!TextUtils.isEmpty(category)) {
+      intent.addCategory(category);
+    }
+    if (data != null) {
+      intent.setData(data);
+    }
+    if (arguments != null) {
+      intent.putExtras(arguments);
+    }
+    if (!TextUtils.isEmpty(packageName)) {
+      intent.setPackage(packageName);
+      if (intent.resolveActivity(applicationContext.getPackageManager()) == null) {
+        Log.i(TAG, "Cannot resolve explicit intent - ignoring package");
+        intent.setPackage(null);
+      }
+    }
+    if (componentName != null) {
+      intent.setComponent(componentName);
+    }
+
+    Log.v(TAG, "Sending intent " + intent);
+    if (activity != null) {
+      activity.startActivity(intent);
+    } else {
+      intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+      applicationContext.startActivity(intent);
+    }
+  }
+
+  /** Caches the given {@code activity} to use for {@link #send}. */
+  void setActivity(@Nullable Activity activity) {
+    this.activity = activity;
+  }
+
+  /**
+   * Caches the given {@code applicationContext} to use for {@link #send}.
+   *
+   * <p>Also resets the cached {@code activity} to null, since the given activity should no longer
+   * be valid after applicationContext changes.
+   */
+  void setApplicationContext(@Nullable Context applicationContext) {
```

Comment:
How do we force people to set the ApplicationContext first and then Activity?
For example:
I want to set a new application context and a new activity. I have to know to set the application context first before calling set activity.

Reply:
I might be missing some context (no pun intended), but why would we not be able to cache both? The app context is always there. It outlives all Activitys. Then an Activity comes and goes.

Reply:
I was thinking that if the application context changed for some reason we know that any previous activity reference we've set is invalid, because the application context should outlive the activities. But @cyanglaz has a good point that a dev may try to call both and happen to call this first so this is kind of a footgun. I removed the `setActivity(null)` logic from here.

Reply:
Can an application context change? I believe the application context is the `Application` class, which I believe is instantiated exactly one time at process launch. Do you know if that's right?

Reply:
That sounds roughly right to me, but I'm not confident enough to totally rule it changing out. I'm suspicious of what would happen in a multiprocess app.